### PR TITLE
[Frontend]  Support strict mode for tool calling

### PR DESCRIPTION
**Pull request by chaunceyjiang**
Co-authored-by: @cjackal <44624812+cjackal@users.noreply.github.com>

## Purpose
[Frontend] Support strict mode for tool calling

## Test Plan

I tested it locally with Minimax 2.5,Qwen2.5, Qwen3.5, Qwen3.6, Qwen3, and DeepSeek V3.2.

I also tested the tool_choice modes required, auto, and named tool selection, and all of them worked correctly.

## Test Result
```

vllm serve /mnt/data3/models/MiniMax/MiniMax-M2.5 --enable-auto-tool-choice --tool-call-parser minimax_m2 --reasoning-parser minimax_m2  -tp 4 --port 8001
vllm bench serve --port 8001 --model /mnt/data3/models/MiniMax/MiniMax-M2.5\
  --backend openai-chat --endpoint /v1/chat/completions \
  --dataset-name hf \
  --dataset-path gorilla-llm/Berkeley-Function-Calling-Leaderboard \
  --bfcl-categories simple,live_simple,multiple \
  --num-warmups 5   --temperature 0   --percentile-metrics ttft,tpot,itl,e2el   \
  --max-concurrency 8 --num-prompts 500
```
main:
```
============ Serving Benchmark Result ============
Successful requests:                     500       
Failed requests:                         0         
Maximum request concurrency:             8         
Benchmark duration (s):                  128.23    
Total input tokens:                      185991    
Total generated tokens:                  86375     
Request throughput (req/s):              3.90      
Output token throughput (tok/s):         673.62    
Peak output token throughput (tok/s):    671.00    
Peak concurrent requests:                17.00     
Total token throughput (tok/s):          2124.12   
---------------Time to First Token----------------
Mean TTFT (ms):                          37.61     
Median TTFT (ms):                        37.76     
P99 TTFT (ms):                           52.79     
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          11.52     
Median TPOT (ms):                        11.53     
P99 TPOT (ms):                           11.66     
---------------Inter-token Latency----------------
Mean ITL (ms):                           15.41     
Median ITL (ms):                         11.40     
P99 ITL (ms):                            23.12     
----------------End-to-end Latency----------------
Mean E2EL (ms):                          2013.90   
Median E2EL (ms):                        1576.51   
P99 E2EL (ms):                           5956.04 
```

this pr
```
============ Serving Benchmark Result ============
Successful requests:                     500       
Failed requests:                         0         
Maximum request concurrency:             8         
Benchmark duration (s):                  130.70    
Total input tokens:                      185991    
Total generated tokens:                  88166     
Request throughput (req/s):              3.83      
Output token throughput (tok/s):         674.57    
Peak output token throughput (tok/s):    643.00    
Peak concurrent requests:                16.00     
Total token throughput (tok/s):          2097.60   
---------------Time to First Token----------------
Mean TTFT (ms):                          37.79     
Median TTFT (ms):                        37.38     
P99 TTFT (ms):                           69.77     
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          11.52     
Median TPOT (ms):                        11.54     
P99 TPOT (ms):                           11.65     
---------------Inter-token Latency----------------
Mean ITL (ms):                           15.34     
Median ITL (ms):                         11.42     
P99 ITL (ms):                            23.14     
----------------End-to-end Latency----------------
Mean E2EL (ms):                          2056.44   
Median E2EL (ms):                        1646.51   
P99 E2EL (ms):                           5965.42   
==================================================

```
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by cjackal** [comment]
this PR supercedes #43678

**Comment by cjackal** [comment]
`glm_4_7` (GLM-4.7 / GLM-5) also works well in our internal test.

File: vllm/envs.py [hunk 1]
```diff
@@ -239,7 +239,6 @@
     VLLM_GPT_OSS_HARMONY_SYSTEM_INSTRUCTIONS: bool = False
     VLLM_SYSTEM_START_DATE: str | None = None
     VLLM_TOOL_JSON_ERROR_AUTOMATIC_RETRY: bool = False
-    VLLM_ENFORCE_STRICT_TOOL_CALLING: bool = False
```
Comment (cjackal) [inline vllm/envs.py:230]: Do we want to disallow non-stag usage? We have some FSM advancement failure type blockers (which emits 500 status error on client side) for (xgrammar + PP) & (xgrammar + speculative decoding), so it may hurt UX if we have no way to disable strict tool calling feature.

File: vllm/tool_parsers/abstract_tool_parser.py [hunk 2]
```diff
@@ -170,7 +178,17 @@ def adjust_request(
...
+        if self.structural_tag_model is None:
+            return None
+
+        from vllm.tool_parsers.structural_tag_registry import get_model_structural_tag
+
+        return get_model_structural_tag(
+            model=self.structural_tag_model,
+            tools=request.tools,
+            tool_choice=request.tool_choice,
+            reasoning=False,
```
Comment (sfeng33) [inline vllm/tool_parsers/abstract_tool_parser.py]: We'd want the whole grammar with the reasoning part, but at the level of the tool parser, we don't have this knowledge, which is why I think the tag logic would be better to reside in https://github.com/vllm-project/vllm/blob/main/vllm/parser/abstract_parser.py#L616, so that we can check reasoning kwargs, etc. Wdyt?

Comment (chaunceyjiang) [reply vllm/tool_parsers/abstract_tool_parser.py]: Done.

Currently, get_xgrammar_model_structural_tag() is always called with reasoning=False, which means the reasoning content is not constrained.

The reason is that during my local testing with DSV3.2, I observed that enabling constraints on the reasoning phase could occasionally cause the model to get stuck in a generation loop. I plan to investigate the root cause and will re-enable this parameter once I have a better understanding of the issue.

File: vllm/envs.py (hunk 1 above)
Comment (chaunceyjiang) [reply vllm/envs.py:230]: Done.

File: vllm/tool_parsers/abstract_tool_parser.py (hunk 2 above)
Comment (cjackal) [reply vllm/tool_parsers/abstract_tool_parser.py]: > The reason is that during my local testing with DSV3.2, I observed that enabling constraints on the reasoning phase could occasionally cause the model to get stuck in a generation loop. I plan to investigate the root cause and will re-enable this parameter once I have a better understanding of the issue.

I also got a similar behavior for qwen3.5, one possible explanation is that the builtin structural tags mask non-end-of-reasoning special tokens but it is possible that the unconstraint logits want to escape reasoning phase by non-end-of-reasoning tokens (e.g.   
tool_start), and in that case the model lost the chance to escape the reasoning phase and malfunction. I haven't dug deep though.

File: vllm/parser/abstract_parser.py [hunk 3]
```diff
@@ -618,10 +619,47 @@ def adjust_request(
...
+            request = self._apply_structural_tag(request)
         if self._tool_parser is not None:
             request = self._tool_parser.adjust_request(request)
         return request
 
+    def _apply_structural_tag(
+        self, request: ChatCompletionRequest | ResponsesRequest
+    ) -> ChatCompletionRequest | ResponsesRequest:
+        if (
+            not isinstance(request, ChatCompletionRequest)
```
Comment (sfeng33) [inline vllm/parser/abstract_parser.py:441]: Do you want to leave the response api part as follow up?

**Review by sfeng33 — APPROVED** [review]
Thank you both for the testing, LGTM!
I added @cjackal as co-author.

File: vllm/tool_parsers/abstract_tool_parser.py (hunk 2 above)
Comment (chaunceyjiang) [reply vllm/tool_parsers/abstract_tool_parser.py]: > I also got a similar behavior for qwen3.5, one possible explanation is that the builtin structural tags mask non-end-of-reasoning special tokens but it is possible that the unconstraint logits want to escape reasoning phase by non-end-of-reasoning tokens (e.g.
tool_start), and in that case the model lost the chance to escape the reasoning phase and malfunction.

Yeah, that's definitely possible. We've seen this before with a few tool parsers , sometimes the model exits the reasoning phase by emitting a tool-related token instead of a dedicated end-of-reasoning token.

I haven't dug into it deeply yet either.

That said, just constraining the tool-calling phase already seems to help a lot. For example, when I was testing MiniMax 2.5 locally, tool calling would start to break down after 200+ turns in a conversation. With the structural-tag constraints on tool calling, that issue seems to be gone.

File: vllm/parser/abstract_parser.py (hunk 3 above)
Comment (chaunceyjiang) [reply vllm/parser/abstract_parser.py:441]: Yeah, I'll add support for this in the next PR.

Comment (sfeng33) [reply vllm/parser/abstract_parser.py:441]: Sounds good, thanks!

**Comment by ankrovv** [comment]
I tested this with GPT-OSS 120B using:

```bash
VLLM_ENFORCE_STRICT_TOOL_CALLING=true
--enable-auto-tool-choice
--tool-call-parser openai
--reasoning-parser openai_gptoss
```

I noticed the structural tag is not applied in the live GPT-OSS chat render path for tool_choice="required".  The request for gpt goes through the  _make_request_with_harmony() path which bypasses the generic preprocess_chat(... parser=self.parser ...) path where the structural_tag gets applied

And, if I force the generated Harmony structural tag, it leaks raw Harmony markers into assistant content (reproducible when using ambiguous prompts).
For prompt "Hi, how are you?" with tool_choice="required":

```
message.content = <|channel|>commentary to=functions.get_weather<|constrain|>json<|message|>{...}
tool_calls = []
finish_reason = stop
```

Also, the generated Harmony tag for tool_choice="required" contains: "at_least_one": false so even if wired into the live path, it does not appear to enforce required-tool semantics. Could you look into this? Personally, I think using json constraints for gpt oss is a lot simpler since it avoids the overall harmony leak but feel free to do your own deep-dive @chaunceyjiang @cjackal

**Comment by chaunceyjiang** [comment]
@ankrovv Yeah, GPT-OSS isn't supported yet. As you mentioned, it currently goes through a separate code path.

We're in the process of unifying the Harmony and non-Harmony paths, and once that's done, we'll integrate it with structural_tags harmony tags as well.

Overall, using structural tags has been quite effective at improving tool-calling accuracy and reliability.

**Comment by ywang96** [comment]
Can we resolve the merge conflict so that we can get this one in? Thanks for the great work from everyone!

**Comment by yzong-rh** [comment]
There may be some incompatibilies btw xgrammar's structural tags and vLLM handling. For example, for MiniMax, the xgrammar tag is:
```
{
  "type": "structural_tag",
  "format": {
    "type": "sequence",
    "elements": [
      {
        "type": "const_string",
        "value": "\n</think>\n\n"
      },
      {
        "type": "const_string",
        "value": "\n\n"
      },
      ...
        ]
      }
    ]
  }
}
```

Note that it forces the `\n</think>\n\n` and `\n\n` to be generated.

However, our MinimaxM2 reasoning parser https://github.com/vllm-project/vllm/blob/6f573f486bc659adf51a8d4639e225097f2d8d39/vllm/reasoning/minimax_m2_reasoning_parser.py#L22 only considers `is_reasoning_ended` (and allow structural tags to be applied) when the `</think>` token is already generated. This forces the model to generate an extra `</think>`.

I served a model locally with
```
VLLM_LOGGING_LEVEL=DEBUG vllm serve MiniMaxAI/MiniMax-M2.7 \
  --tensor-parallel-size 4 \
  --tool-call-parser minimax_m2 \
  --reasoning-parser minimax_m2  \
  --enable-auto-tool-choice \
  --enforce-eager \
  --trust-remote-code \
  --enable-log-requests \
  --enable-log-outputs | tee vllm-serve-minimax.log
```
then sent a simple request. I inspected `vllm-serve-minimax.log` to get the actual generated token IDs.
```
...
]~b]user
What's NVIDIA's current P/E ratio?[e~[
]~b]ai
<think>
The user is asking about NVIDIA's current P/E ratio. I need to use the get_current_pe_ratio function with the ticker "NVDA" for NVIDIA.
</think>
</think>



<minimax:tool_call>
<invoke name="get_current_pe_ratio">
<parameter name="ticker">NVDA</parameter>
</invoke>
</minimax:tool_call>[e~[
```

Note the duplicate `</think>`. This does not occur on main.

**Comment by yzong-rh** [comment]
The same problem is worse for Harmony. https://github.com/vllm-project/vllm/blob/6f573f486bc659adf51a8d4639e225097f2d8d39/vllm/reasoning/gptoss_reasoning_parser.py#L85
`GptOssReasoningParser` doesn't consider reasoning ended until `<|channel|>final ... <|message|>` is seen. 

- If the model does generate tool call in the harmony format (`<|channel|>commentary to=functions. ...`), the structural tags won't ever activate. Strict tool call isn't enforced.
- If the model doesn't make tool call and instead generates the final output, the structural tags will kick in and force it to generate a wrongly formattted tool call instead (`<|channel|>final<|message|><|channel|>commentary to=functions. ...`)

> Yeah, GPT-OSS isn't supported yet. As you mentioned, it currently goes through a separate code path.

I recently refactored Harmony to the unified Parser path for Chat Completions: https://github.com/vllm-project/vllm/pull/45171 and https://github.com/vllm-project/vllm/pull/45104. It should now receive structural tags (on the Chat Completions path) after the rebase, leading to the above issues.

For now, a simple fix could be to not enable structural tag / strict tool calling for GPT-OSS (which would also fix the merge conflict) and validate (compare generated tokens with and without structural tags) other models to ensure they are correct. (Happy to help on that front, qwen3 seems to be correct from a prelim check).

cc @bbrowning who has encountered similar issues with constrained strict tool calling.

**Comment by chaunceyjiang** [comment]
@yzong-rh Thanks for catching this so carefully.

Regarding the MiniMax issue, I believe the root cause is in xgrammar's built-in `structural_tag` implementation (https://github.com/mlc-ai/xgrammar/blob/main/python/xgrammar/builtin_structural_tag.py#L1455-L1456). I'll discuss it with their team later. In the meantime, I've reimplemented that part on our side, and the issue has been resolved.

As for the OpenAI/Harmony issue, the previous Harmony path bypassed `structural_tag` entirely. Since I hadn't tested that path yet, and because of your refactoring there, I haven't had a chance to verify the behavior with the new implementation. For now, I've temporarily disabled the OpenAI parser handling until I can properly test it.

**Comment by yzong-rh** [comment]
> I believe the root cause is in xgrammar's built-in structural_tag implementation (https://github.com/mlc-ai/xgrammar/blob/main/python/xgrammar/builtin_structural_tag.py#L1455-L1456). I'll discuss it with their team later.

Great, thank you. I went through the structural tags in `XGRAMMAR_BUILTIN_STRUCTURAL_TAG_MODELS`. `minimax` was indeed the only implementation with that behavior.

A remaining risk is `kimi`, whose `is_reasoning_end` allows "implicit reasoning end via tool call section".
https://github.com/vllm-project/vllm/blob/7021be66e8c351fa819fd07b4053e946f11c1147/vllm/reasoning/kimi_k2_reasoning_parser.py#L76-L83

When that triggers, the start of a tool section is already generated. `xgrammar`'s tags would force the start of tool section to be generated again. But this only occurs if kimi "forgot" to close `</think>`. So not an `xgrammar` problem, and not a blocker.

> For now, I've temporarily disabled the OpenAI parser handling until I can properly test it.

Sg. There might be some challenges due to the harmony channel format. Happy to look into it if you don't have the bandwidth.

**Comment by bbrowning** [comment]
Since this merged I'm seeing Qwen 3.6 models getting stuck in infinite whitespace generation loops regularly in relatively simple tool calling scenarios with our out of the box setup. It's easily triggered with things like BFCL multi_turn_base, resulting in extremely long generation times and timeouts while the model generates thousands of newline tokens per turn.

We've flip this on by default for a lot of models now in tool calling scenarios, and at least in some cases this has made things worse. We may need a more selective testing and case-by-case enabling of this.
